### PR TITLE
Support Redis Cluster

### PR DESCRIPTION
**Pull request by gnought**

### CODE REVIEW

**Comment by mcollina** [comment]
CI is failing, would you mind fixing that? Do you think it would be possible to add a unit test for this feature?

**Comment by gnought** [comment]
Unit tests are added. Finally all good now. Please have a review and push into the mainstream.

File: test.js [hunk 1]
```diff
@@ -1,83 +1,36 @@
...
-    t.end()
+  require('./redisTest.js')({
+    builder: builder,
+    isCluster: isCluster || false,
+    test: test
   })
-})
+}
+
+runTests(require('./'))
```
Comment (mcollina) [inline test.js]: can you move this require to the top of the file, and also the next one?

File: mqemitter-redis.js [hunk 2]
```diff
@@ -74,20 +85,26 @@ inherits(MQEmitterRedis, MQEmitter)
 })
 
 MQEmitterRedis.prototype.close = function (cb) {
-  var count = 2
   var that = this
 
-  function onEnd () {
-    if (--count === 0) {
-      that._close(cb)
+  setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (gnought) [reply mqemitter-redis.js]: Most tests, e.g. 'support on and emit' test
```
    e.on('hello world', function (message, cb) {
      t.equal(e.current, 1, 'number of current messages')
      t.deepEqual(message, expected)
      t.equal(this, e)
      cb()
    }, function () {
      e.emit(expected, function () {
        e.close(function () {
          t.pass('closed')
        })
      })
    })
```
we listen the subscribed topic `hello world`, then a message is emitted to callback to ensure they are in order. It's fine and the event will be triggered after there is a message in 'hello world' topic comes.
However we call close in emit callback, redis/redis cluster will close all connections before we get any subscription call. The test plan will get 1 from `t.pass('close')`, but not 4.
I think close() is an infrequent call and we want to close the redis connection if we really want to. 
Make a delay in the close() call is a way unless we want to change all test cases.

[the remark this answers is not in the document]
Comment (gnought) [reply mqemitter-redis.js]: It is a race condition. Sometimes redis could deliver all subscriptions before the close(), but not always true

File: mqemitter-redis.js [hunk 3]
```diff
@@ -74,20 +85,26 @@ inherits(MQEmitterRedis, MQEmitter)
...
+    var handleClose = function () {
+      Promise.all([
+        that.subConn.quit().then(() => { that.subConn.disconnect() }),
+        that.pubConn.quit().then(() => { that.pubConn.disconnect() })
+      ]).then(() => { that._close(cb) })
```
[the remark this answers is not in the document]
Comment (gnought) [reply mqemitter-redis.js]: yes, added catch in 4ae07fba8c0cdc226d10ad6fc47d544c664e26a2
changed to callback in 006ce0eb1747ddf3f58e589450f4adb5d0872f2d

File: test.js (hunk 1 above)
Comment (gnought) [reply test.js]: sure. here you are ab5956f6ae2736980a9b748bb0ea3043344c372c

File: mqemitter-redis.js [hunk 4]
```diff
@@ -51,15 +62,15 @@ function MQEmitterRedis (opts) {
   })
 
   this.subConn.on('error', function (err) {
-    that.state.emit('error', err)
+    that.state.emit('subError', err)
```
[the remark this answers is not in the document]
Comment (gnought) [reply mqemitter-redis.js]:
```
test('ioredis error event', function (t) {
  var e = redis({ host: '127' })

  t.plan(1)

  e.state.once('error', function (err) {
    t.deepEqual(err.message.substr(0, 7), 'connect')
    e.close(function () {
      t.end()
    })
  })
})
```
In the original version, once we get an error from either publisher or subscriber, `close()` is triggered and no more error event be triggered after `t.end()`.

As there is a race condition mentioned in https://github.com/mcollina/mqemitter-redis/pull/13#discussion_r298803764, we delay the `close()` a bit. In this way, the other error event either from publisher or subscriber will come before the first `t.end()` is called. The error will be uncaught and fail the test.

Separate the error is a way, or you have a better idea?

File: mqemitter-redis.js [hunk 5]
```diff
@@ -51,15 +62,15 @@ function MQEmitterRedis (opts) {
...
   })
 
   this.pubConn.on('connect', function () {
     that.state.emit('pubConnect')
   })
 
   this.pubConn.on('error', function (err) {
-    that.state.emit('error', err)
+    that.state.emit('pubError', err)
```
[the remark this answers is not in the document]
Comment (gnought) [reply mqemitter-redis.js]: Same as above.

**Comment by gnought** [comment]
I will resubmit PR with better handling

**Comment by r3na** [comment]
will this PR ever come back?

**Pull request closed without merging**